src/util/small_map.h

```diff
@@ -0,0 +1,560 @@
+#include <cstdint>
+#include <cstring>
+#include <limits>
+#include <memory>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+#include "invariant.h"
+
+// When we don't have constexpr
+#if !(defined(__GNUC__) || _MSC_VER >= 1900)
+
+template <std::size_t N>
+struct num_bitst
+{
+  static const std::size_t value = 1 + num_bitst<(N >> 1)>::value;
+};
+
+template <>
+struct num_bitst<1>
+{
+  static const std::size_t value = 1;
+};
+
+template <typename T, std::size_t B, typename U = std::integral_constant<T, 1>>
+struct indicator_maskt
+{
+  static const T value =
+    U::value |
+    indicator_maskt<T, B, std::integral_constant<T, (U::value << B)>>::value;
+};
+
+template <typename T, std::size_t B>
+struct indicator_maskt<T, B, std::integral_constant<T, 0>>
+{
+  static const T value = 0;
+};
+
+#endif
+
+/// Map from small integers to values
+///
+/// A data structure that maps small integers (in {0, ..., Num-1}) to values.
+/// It is designed to be more memory-efficient than std::map for this use case.
+///
+/// In the following we give some data about the memory consumption of
+/// small_mapt compared to std::map, on a 64-bit Linux system with GNU STL. It
+/// was determined with valgrind/massif. We configured the maps as
+/// `small_mapt<uint64_t, uint32_t, Num=8>` and `std::map<uint8_t, uint64_t>`.
+/// Thus, indices to small_mapt can be in {0, ..., 7} for this configuration.
+///
+/// Below we give the memory consumption (in bytes) of small_mapt and std::map,
+/// both for when containing 8 elements, and when being empty. The value for
+/// "useful" indicates the number of bytes requested, the value for "total" also
+/// includes padding bytes.
+///
+/// small_mapt:
+/// - 8 elements:
+///   * useful:  80 B
+///   * total:  120 B
+/// - empty:
+///   * useful:  16 B
+///   * total:   32 B
+///
+/// std::map:
+/// - 8 elements:
+///   * useful: 432 B
+///   * total:  512 B
+/// - empty:
+///   * useful:  48 B
+///   * total:   64 B
+///
+/// \tparam T mapped type
+/// \tparam Ind unsigned integer type, used to map integer indices to internal
+///   indices that index into the memory block that stores the mapped values
+/// \tparam Num gives range of valid indices, i.e., the valid indices are {0,
+///   ..., Num-1}, must satisfy Num * num_bits(Num-1) + Num < sizeof(Ind) * 8,
+///   with num_bits(n) denoting the minimum number of bits required to represent
+///   integer n
+template <typename T, typename Ind = uint32_t, std::size_t Num = 8>
+class small_mapt
+{
+public:
+  static_assert(std::is_unsigned<Ind>::value, "");
+
+  typedef Ind index_fieldt;
+
+  friend void small_map_test();
+
+  small_mapt() : ind(0), p(nullptr)
+  {
+  }
+
+  // Memory
+
+  void copy(T *dest, const T *src, std::size_t n) const
+  {
+    for(std::size_t i = 0; i < n; i++)
+    {
+      new(dest + i) T(*(src + i));
+    }
+  }
+
+  T *allocate(std::size_t n) const
+  {
+    if(n == 0)
+      return nullptr;
+
+    T *mem = (T *)malloc(sizeof(T) * n);
```

Comment:
Why all this code with malloc/realloc instead of just using `std::vector` (which already does all of these things for you)?

Also, this leaves the result of the allocation uninitialised which is probably ok for POD types, but non-POD types need to be initialised with placement-new.